---
 libs/vkd3d/swapchain.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/libs/vkd3d/swapchain.c b/libs/vkd3d/swapchain.c
index 77a83099..c11c7379 100644
--- a/libs/vkd3d/swapchain.c
+++ b/libs/vkd3d/swapchain.c
@@ -1378,7 +1378,7 @@ static HRESULT dxgi_vk_swap_chain_init_sync_objects(struct dxgi_vk_swap_chain *c
     }
     else
     {
-#define DEFAULT_FRAME_LATENCY 3
+#define DEFAULT_FRAME_LATENCY 1
         /* 3 frames is the default in the API, so that's what we should report to application if asks. */
         chain->frame_latency = DEFAULT_FRAME_LATENCY;
     }
@@ -1914,9 +1914,13 @@ static void dxgi_vk_swap_chain_recreate_swapchain_in_present_task(struct dxgi_vk
     if (strlen(count_env) > 0)
     {
         override_image_count = strtoul(count_env, NULL, 0);
-        swapchain_create_info.minImageCount = max(surface_caps.minImageCount, override_image_count);
-        INFO("Overriding swapchain images to %u.\n", swapchain_create_info.minImageCount);
-    }
+        swapchain_create_info.minImageCount = override_image_count; //max(surface_caps.minImageCount, override_image_count);
+    }
+    else
+    {
+        swapchain_create_info.minImageCount = 2;
+    }
+    INFO("Overriding swapchain images to %u.\n", swapchain_create_info.minImageCount);
 
     if (surface_caps.maxImageCount && swapchain_create_info.minImageCount > surface_caps.maxImageCount)
         swapchain_create_info.minImageCount = surface_caps.maxImageCount;
-- 
2.51.0

